File: pact-jvm-consumer/src/main/java/au/com/dius/pact/consumer/dsl/PactDslWithProvider.java
package au.com.dius.pact.consumer.dsl;

import au.com.dius.pact.consumer.ConsumerPactBuilder;
import au.com.dius.pact.model.ProviderState;

import java.util.HashMap;
import java.util.Map;

public class PactDslWithProvider {
    private ConsumerPactBuilder consumerPactBuilder;
    private String providerName;

    public PactDslWithProvider(ConsumerPactBuilder consumerPactBuilder, String provider) {
        this.consumerPactBuilder = consumerPactBuilder;
        this.providerName = provider;
    }

    /**
     * Describe the state the provider needs to be in for the pact test to be verified.
     *
     * @param state Provider state
     */
    public PactDslWithState given(String state) {
        return new PactDslWithState(consumerPactBuilder, consumerPactBuilder.getConsumerName(), providerName,
          new ProviderState(state));
    }

    /**
     * Describe the state the provider needs to be in for the pact test to be verified.
     *
     * @param state Provider state
     * @param params Data parameters for the state
     */
    public PactDslWithState given(String state, Map<String, Object> params) {
        return new PactDslWithState(consumerPactBuilder, consumerPactBuilder.getConsumerName(), providerName,
          new ProviderState(state, params));
    }

    /**
     * Describe the state the provider needs to be in for the pact test to be verified.
     *
     * @param firstKey Key of first parameter element
     * @param firstValue Value of first parameter element
     * @param paramsKeyValuePair Additional parameters in key-value pairs
     */
    public PactDslWithState given(String state, String firstKey, Object firstValue, Object... paramsKeyValuePair) {

        if (paramsKeyValuePair.length % 2 != 0) {
            throw new IllegalArgumentException("Pair key value should be provided, but there is one key without value.");
        }

        final Map<String, Object> params = new HashMap<>();
        params.put(firstKey, firstValue);

        for (int i = 0; i < paramsKeyValuePair.length; i+=2) {
            params.put(paramsKeyValuePair[i].toString(), paramsKeyValuePair[i+1]);
        }

        return new PactDslWithState(consumerPactBuilder, consumerPactBuilder.getConsumerName(), providerName,
                new ProviderState(state, params));
    }

    /**
     * Description of the request that is expected to be received
     *
     * @param description request description
     */
    public PactDslRequestWithoutPath uponReceiving(String description) {
        return new PactDslWithState(consumerPactBuilder, consumerPactBuilder.getConsumerName(), providerName)
                .uponReceiving(description);
    }

}

File: consumer/src/main/kotlin/au/com/dius/pact/consumer/dsl/PactDslWithProvider.kt
package au.com.dius.pact.consumer.dsl

import au.com.dius.pact.consumer.ConsumerPactBuilder
import au.com.dius.pact.core.model.PactSpecVersion
import au.com.dius.pact.core.model.ProviderState
import au.com.dius.pact.core.support.json.JsonValue

open class PactDslWithProvider @JvmOverloads constructor(
  val consumerPactBuilder: ConsumerPactBuilder,
  private val providerName: String,
  val version: PactSpecVersion = PactSpecVersion.V3
) {
  private var defaultRequestValues: PactDslRequestWithoutPath? = null
  private var defaultResponseValues: PactDslResponse? = null
  private val additionalMetadata: MutableMap<String, JsonValue> = mutableMapOf()

  /**
   * Describe the state the provider needs to be in for the pact test to be verified.
   *
   * @param state Provider state
   */
  fun given(state: String): PactDslWithState {
    return PactDslWithState(consumerPactBuilder, consumerPactBuilder.consumerName, providerName,
      ProviderState(state), defaultRequestValues, defaultResponseValues, version, additionalMetadata)
  }

  /**
   * Describe the state the provider needs to be in for the pact test to be verified.
   *
   * @param state Provider state
   * @param params Data parameters for the state
   */
  fun given(state: String, params: Map<String, Any?>): PactDslWithState {
    return PactDslWithState(consumerPactBuilder, consumerPactBuilder.consumerName, providerName,
      ProviderState(state, params), defaultRequestValues, defaultResponseValues, version, additionalMetadata)
  }

  /**
   * Describe the state the provider needs to be in for the pact test to be verified.
   *
   * @param firstKey Key of first parameter element
   * @param firstValue Value of first parameter element
   * @param paramsKeyValuePair Additional parameters in key-value pairs
   */
  fun given(state: String, firstKey: String, firstValue: Any?, vararg paramsKeyValuePair: Any): PactDslWithState {
    require(paramsKeyValuePair.size % 2 == 0) {
      "Pair key value should be provided, but there is one key without value."
    }
    val params = mutableMapOf(firstKey to firstValue)
    var i = 0
    while (i < paramsKeyValuePair.size) {
      params[paramsKeyValuePair[i].toString()] = paramsKeyValuePair[i + 1]
      i += 2
    }
    return PactDslWithState(consumerPactBuilder, consumerPactBuilder.consumerName, providerName,
      ProviderState(state, params), defaultRequestValues, defaultResponseValues, version, additionalMetadata)
  }

  /**
   * Description of the request that is expected to be received
   *
   * @param description request description
   */
  fun uponReceiving(description: String): PactDslRequestWithoutPath {
    return PactDslWithState(consumerPactBuilder, consumerPactBuilder.consumerName, providerName,
      defaultRequestValues, defaultResponseValues, version, additionalMetadata)
      .uponReceiving(description)
  }

  fun setDefaultRequestValues(defaultRequestValues: PactDslRequestWithoutPath) {
    this.defaultRequestValues = defaultRequestValues
  }

  fun setDefaultResponseValues(defaultResponseValues: PactDslResponse) {
    this.defaultResponseValues = defaultResponseValues
  }

  /**
   * Adds additional values to the metadata section of the Pact file
   */
  fun addMetadataValue(key: String, value: String): PactDslWithProvider {
    additionalMetadata[key] = JsonValue.StringValue(value)
    return this
  }

  /**
   * Adds additional values to the metadata section of the Pact file
   */
  fun addMetadataValue(key: String, value: JsonValue): PactDslWithProvider {
    additionalMetadata[key] = value
    return this
  }
}
